### ISSUE TITLE
DeepLift / DeepLiftShap incorrect with common ReLU usage

### ISSUE THREAD

**Issue by jmschrei** [meta-pytorch/captum#1082]
Hello all.

I have a model with the following style:

```python
class Test1(torch.nn.Module):
    def __init__(self....):
        self.conv1 = torch.nn.Conv1d(...)
        self.conv2 = torch.nn.Conv2d(...)

    def forward(self, X):
        X = torch.nn.ReLU()(self.conv1(X))
        X = torch.nn.ReLU()(self.conv2(X))
        return X
```

This will run with DeepLift and DeepLiftShap but it will not produce the same attributions as an identical model that yields predictions within machine precision (1e-14) implemented in Keras and inspected using the TFDeepExplainer in the shap repo, even when the baselines are held constant.

![image](https://user-images.githubusercontent.com/3916816/205816042-435414d9-2706-4e84-9ddc-f8b23b152a49.png)

Naturally, if I create one ReLU instance and use it multiple times in the forward function I get issues. However, if I create one ReLU object per layer I need to call it on and use each exactly once in the forward function, I get the exact same attributions as TFDeepExplainer, e.g.:

```python
class Test2(torch.nn.Module):
    def __init__(self....):
        self.conv1 = torch.nn.Conv1d(...)
        self.conv2 = torch.nn.Conv2d(...)
        self.relu1 = torch.nn.ReLU()
        self.relu2 = torch.nn.ReLU()

    def forward(self, X):
        X = self.relu1(self.conv1(X))
        X = self.relu2(self.conv2(X))
        return X
```

![image](https://user-images.githubusercontent.com/3916816/205816481-ecad2a02-3f44-431d-b52b-ab415ef2323a.png)

I only figured this out after three days of trying to debug the differences in the packages. Is it possible to add this quirk much more prominently to the documentation of DeepLift/DeepLiftShap? Given that these seemingly similar model implementations produce different results, people might be interested in knowing. 

Relatedly, do you know why this is the case? I'm sure it has to do with the hooks that are being setup, but I couldn't figure it out.

Thanks!

**Comment by Vejni** [comment]
It would also be nice to know if this behaviour is reproduced with sequential blocks or not.

**Comment by NarineK** [comment]
@jmschrei, I think that the issue might be the following:
In PyTorch we use hooks on the modules such as on ReLUs to apply the rules. Unfortunately, since PyTorch hooks are not designed to be graph-based, we don't know where is the hook called from and in which order. What ends up happening is that the last call overrides previous results. We had ideas of counting and keeping track of the order in which the hooks are called. This way we could avoid that issue but in order to get accurate results it is better to use each ReLU only once in the computation graph. 

cc: @vivekmig

**Comment by jmschrei** [comment]
Hi @NarineK. That makes sense to me. Would it be possible to make that explanation more explicit in the documentation for DeepLift / DeepLiftShap? Given that it doesn't raise an error when you do it wrong, it seems worth being as clear as possible.

**Comment by NarineK** [comment]
Yes, that's a good point. We can create a PR to add that explanation. Also, feel free to create a PR if you'd like to contribute it to Captum.

**Comment by craymichael** [comment]
Hi, this restriction is now documented in the FAQ under functional non-linearities and reused modules: https://captum.ai/docs/faq#can-my-model-use-functional-non-linearities-eg-nnfunctionalrelu-or-can-reused-modules-be-used-with-captum. DeepLift / DeepLiftShap rely on hooks for supported nonlinear modules, so functional activations and reused module instances can produce incorrect multipliers. Closing this since the documentation now calls out the limitation.
